### ISSUE TITLE
Improve fetch/lint/deploy workflow

### ISSUE THREAD

**Issue by sebamarynissen** [sebamarynissen/simtropolis-channel#12]
@memo33 I'm currently struggling with the actions. I have managed to automatically create a PR for every package that gets added, but the problem is that the lint action won't run on that PR, because GitHub refuses to run actions on commits made by a bot.

I've already tried committing with my own identity (through a PAT) instead of using the github-bot token, but to no avail. Do you have any experience with this?

I'm now trying another approach where I manually trigger the lint workflow, let's see if that works, but it feels hacky.

**Comment by sebamarynissen** [comment]
There might be a way to do this with [GitHub Apps](https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/about-creating-github-apps) apparently. I'm going to experiment with it.

**Comment by memo33** [comment]
I'm not too experienced with this, unfortunately. Though, in your triggering `fetch` workflow, couldn't you just add a second job that is executed after the first one completes? This would ensure that everything runs as part of the same event trigger. Something like:
```yaml
jobs:
  add-from-simtropolis:
    …

  deploy:
    needs: add-from-simtropolis
    uses: memo33/sc4pac-actions/.github/workflows/sc4pac-channel.yaml@main
    with:
      …
```

Calling the `lint` action directly should be fine as well. Just make sure to define the input parameter (not sure why it doesn't seem to error out for you):
```yaml
      uses: memo33/sc4pac-actions/actions/lint@main
      with:
        path: src/yaml
```
Inlining the deploy job would be more complicated and I'd avoid it if possible. Though, maybe I could refactor it so you could call just that job if necessary.

**Comment by sebamarynissen** [comment]
Well, the thing is that the fetch action and the linting action should run off different targets. The fetch action runs off `main`, and it can create multiple PRs - for example when multiple packages have been added since the last call to the api. This workflow allows that 1 package with invalid metadata doesn't block the othere one with valid metadata from being merged.

After the PRs have been created, the linting action should run for every PR, and when it succeeds it should merge the PR automatically, which then in turn triggers a deploy. However, I think your approach of running everything as part of the same event trigger might still work if I provide an output from my fetch action containing all the created branches and PRs, and then use the `matrix` option to run the linting for every one of them. I think that should work.

**Comment by memo33** [comment]
Ah, I see. Hm. The `matrix` approach is not ideal I think. You'd want all the PRs to be processed _sequentially_ one by one, not in parallel. This is to ensure that the linting runs on the PR's merge commit that reflects the future state of the PR if it gets merged. Otherwise, two different PRs could introduce conflicts, even if the actions for each PR passed beforehand.

This sequential processing would be desirable in general I think, so it makes sense that the actions don't eagerly run on every opened PR directly.

I imagine it would all have to be part of/triggered by the `fetch` workflow that opened the PRs. Though, I don't see a way to do that by re-using the actions I defined. One would somehow have to call the underlying scripts directly.

**Comment by sebamarynissen** [comment]
Yeah, I see your point with the `matrix` approach, although I got it more or less working now.

The easiest solution for me would be to just put everything in a huge JavaScript action where I handle all the PR stuff with the GitHub api. I'm already doing this now pretty much. That way, I could generate the PRs, lint them and perhaps merge them one by one easily. The only thing that's blocking me though is that I need a way to run your linting from within JavaScript in that case.

Perhaps that there's a different option too though. If GitHub has an option for the matrix to run sequentially rather than in parallel, then I could do an "update from main" before linting every PR. In theory this could introduce merge conflicts, but as files from different packages normally shouldn't collide - and if they do, I could probably prevent this in my fetch action - so that isn't much of a problem.

**Comment by sebamarynissen** [comment]
New workflow that I think might work:

- Job `fetch`: checks out the `main` branch, calls the STEX api and generates all the necessary metadata files. It *does not* commit them to the repo. It sets the following array as its output:
```json
[{
  "branch": "package/group-name",
  "files": [
    "src/yaml/group/name.yaml"
  ],
  "body": "Body text for PR..."
}]
```
- Job `commit`: 
  - Step `create-prs`: runs a *sequential* matrix strategy - can be done [by setting `max-parallel: 1`](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow#defining-the-maximum-number-of-concurrent-jobs). Each job runs a custom JavaScript action that takes the generated files, commits them and then creates a PR if it does not exists yet. If the PR already exists, then it will first be updated from the main branch.
  - Step `lint`: runs the linting action
  - Step `deploy`: merges into main and deploys to gh pages if the linting action is successful.

**Comment by sebamarynissen** [comment]
I finally have something fully working (247f1787cea52b8181a911a96b4332fed0ba82a4). I still plan to change it to the workflow above though, but I first need a break from GH actions, lol.

**Comment by memo33** [comment]
> The easiest solution for me would be to just put everything in a huge JavaScript action where I handle all the PR stuff with the GitHub api. I'm already doing this now pretty much. That way, I could generate the PRs, lint them and perhaps merge them one by one easily. The only thing that's blocking me though is that I need a way to run your linting from within JavaScript in that case.

You could call Python from your Javascript code though, couldn't you? Then you could just run the `lint.py` file directly. I think the full command would be the following, but I haven't tested it yet.
```sh
python /home/runner/work/_actions/memo33/sc4pac-actions/main/src/lint.py src/yaml
```
I could also add the script to the PATH for convenience.

Even with parallelism 1, the matrix approach seems a bit awkward, as setting up all those individual runners isn't really needed. There's a maximum of 256 runners, which admittedly is not a limit we'd hit, but hints at a scalability problem.

Instead, my idea would be the following:

- job 1: fetch relevant data from ST
- job 2:
  - step: install the linter via `uses: memo33/sc4pac-actions/actions/lint@main` and run it on `main` to verify you start from a clean slate
  - step: run your Javascript code to sequentially process each relevant PR:
    - commit (if not done in job 1 already), create a provisional merge commit
    - run `lint.py` directly (on the merge commit)
    - if it passes, merge the PR
    - repeat
 - job 3: deploy once via `uses: memo33/sc4pac-actions/.github/workflows/sc4pac-channel.yaml@main`

I haven't fully grasped yet what your current working code is doing, so YMMV.

**Comment by sebamarynissen** [comment]
Yeah I think that might work. I can indeed run python from within JavaScript, but I was thinking that this defeats the "modularity" of all the actions. I'd also have to setup python manually with an action before the JS runs because I can't directly use your action anymore. However, the workflow is going to be so heavily customized anyway that I should probably let go of this idea of modularity and just go with something that works.

Another reason why I wanted to use existing actions as much as possible is that if the linting fails, the creator could go look in the action easily for the errors. However, it's probably better to have the actions bot post a comment in the PR with the result of the linting if it fails as that's way more user friendly. Most creators would probably just get scared of the actions output.

I don't know how good your understanding of JS is, but for the moment the code is still a big mess. That's always the case when prototyping stuff, which JS is ideal for as it lets you do just about anything, haha. Anyway, once the dust settles a bit I plan to still clean it up a little. I have a few integration tests in place already, so refactoring should be easily doable without breaking stuff.

**Comment by memo33** [comment]
> I'd also have to setup python manually with an action before the JS runs because I can't directly use your action anymore.

I think as long as you use the `lint` action regularly beforehand, within the same job, Python and the sc4pac-actions code should still be available. I'm just not sure if Node actions are more encapsulated in that regard.

> Another reason why I wanted to use existing actions as much as possible is that if the linting fails, the creator could go look in the action easily for the errors.

It would be nice to have the log of the workflow run somehow be associated to the commit on the branch of the pull request, so that the log can easily be inspected by viewing the PR. I'm not sure if that's possible. With your current setup, it looks like this is the case, in fact.

Independently of that, posting the lint output in a comment would be very useful, as I imagine most people don't even have a GH account, which is needed to view the actions log.

**Comment by sebamarynissen** [comment]
> It would be nice to have the log of the workflow run somehow be associated to the commit on the branch of the pull request, so that the log can easily be inspected by viewing the PR (the green check mark). I'm not sure if that's possible. With your current setup, it looks like this is the case, in fact.

Yes, I found out that this can be done by using the GitHub api. You can add commit statuses to specific commits, which is what I do now. Before the linting action, I set the status to "pending", and then when it either fails or succeeds I update it again. The workflow also attaches itself as a referencen which means you can see the log by clicking on it. The code that does this is

```js
await github.rest.repos.createCommitStatus({
  owner: context.repo.owner,
  repo: context.repo.repo,
  sha: '${{ matrix.pr.sha }}',
  state: 'pending',
  context: 'repository_dispatch',
  description: 'Running lint',
  target_url: `https://github.com/${context.repo.owner}/${context.repo.repo}/actions/runs/${process.env.GITHUB_RUN_ID}`
});
```

The `target_url` is the reference to the current run of the linter.

> Independently of that, posting the lint output in a comment would be very useful, as I imagine most people don't even have a GH account, which is needed to view the actions log.

Again, reducing the friction as much as possible for ST admins & content creators is the goal, so requiring them to inspect the actions log isn't desirable. Posting the linting results in the PR is a first step, but I would also like to send a DM or something on Simtropolis in a later stadium. That way, creators wouldn't even have to leave Simtropolis and get immediate feedback on whether their package was successfully added or not. To be determined whether we only send a DM when the lint fails, or also when the package has been processed successfully. Someting like

> Hello ${{ package.info.author }}!
>
> This is an auto-generated message to let you know that your sc4pac package ${{ package.group }}:${{ package.name }} is now available on the Simtropolis channel.

**Comment by memo33** [comment]
That's awesome.

**Comment by sebamarynissen** [comment]
Alright, I created a repo for testing the proposed flow, and looks like it's nicely working now. This is an [example](https://github.com/sebamarynissen/st-actions-test/pull/32) of a package that imlediately validates and gets merged, and [this](https://github.com/sebamarynissen/st-actions-test/pull/30) is one that needs fixing, and gets merged after this is done.

I will try to port it to this repo tomorrow.

**Comment by memo33** [comment]
Fantastic. I'm really excited about this. Let me know when/if I should review something.

I've created a new action [memo33/sc4pac-actions/actions/setup-lint@main](https://github.com/memo33/sc4pac-actions/blob/main/actions/setup-lint/action.yaml) to merely _install_ the linter and its Python dependencies. You can subsequently use it directly by calling `sc4pac-lint src/yaml`.

**Comment by sebamarynissen** [comment]
Yeah, me too. I think the number one reason why people quit/don't come back to SC4 is the difficulty to manually compile a decent plugins folder. Let's hope all the work we're doing changes this and attracts a bunch of old and new players.

Speaking of, I think that it would be useful to add some kind of "export/import" function to the sc4pac gui so that people can easily share their plugins folder. Simtropolis could then also include a few of these configurations in a starter guide as an opinionated set of "base plugins" for new players.

**Comment by sebamarynissen** [comment]
It's done. I've ported the workflow over from https://github.com/sebamarynissen/st-actions-test and it seems to work. I also tested the integration with ST. I modified one of my older uploads to include an invalid metadata file, then ran the action, fixed the metadata & uploaded again, then ran the action again. You can see the result in #28 and it's properly available to install with sc4pac-gui. Sweet.

**Comment by memo33** [comment]
Very nice.

Perhaps you could remove this part of the `setup-sc4pac` action. You don't need it for linting. In fact, you could remove the entire action and just call `setup-lint` in the corresponding workflow. If you really want to install the sc4pac executable, use `memo33/sc4pac-actions/actions/setup` instead.
https://github.com/sebamarynissen/simtropolis-channel/blob/c9bfd5524406b21ae9487b891c029d1390e0d5c2/actions/setup-sc4pac/action.yaml#L7-L18

**Comment by sebamarynissen** [comment]
Ah, I thought that would be needed for the deployment. Could you verify that my deploy workflow will work without the sc4pac executable being present?

**Comment by memo33** [comment]
I'm pretty sure it would work. For example, you can see here in this run that sc4pac gets installed as part of the deploy workflow: https://github.com/sebamarynissen/simtropolis-channel/actions/runs/12457069549/job/34771219162#step:3:40

**Comment by sebamarynissen** [comment]
Seems to [have worked](https://github.com/sebamarynissen/simtropolis-channel/actions/runs/12457249636) indeed. Thanks!
